moped-etl/arcgis/components_to_agol.py

```diff
@@ -23,6 +26,63 @@
 )
 
 
+def has_html_tags(html_string_to_check):
+    """Identify if a string contains HTML tags.
+
+    See: https://pypi.org/project/beautifulsoup4/
+
+    Args:
+        html_string_to_check (str): A string to test for HTML tags
+
+    Returns:
+        bool: True if the string contains HTML tags, False otherwise
+    """
+    return bool(BeautifulSoup(html_string_to_check, "html.parser").find())
+
+
+def is_valid_HTML_tag(html_string_to_check):
+    """Identify if a string contains valid HTML.
+
+    See: https://pypi.org/project/beautifulsoup4/
+
+    Args:
+        html_string_to_check (str): A string that contains HTML to test for valid HTML.
+
+    Returns:
+        bool: True if the string contains valid HTML, False otherwise
+    """
+    soup = BeautifulSoup(html_string_to_check, "html.parser")
+    return html_string_to_check == str(soup)
+
+
+def handle_status_updates(features):
+    """Check project status updates for valid or invalid HTML; escape HTML if needed.
+    Project status updates can be plain text, valid HTML, or invalid HTML. If invalid HTML is found,
+    the content of the update is escaped to prevent it from being rejected by AGOL (504 or 400 error).
+
+    Args:
+        features (list): list of Esri feature objects
+
+    Returns:
+        list: list of Esri feature objects with status update HTML escaped if needed
+    """
+    for record in features:
+        id = record["attributes"]["project_id"]
+
+        status_update = record["attributes"]["project_status_update"]
+
+        if status_update != None and has_html_tags(status_update):
+            if not is_valid_HTML_tag(status_update):
+                logger.info(
+                    f"Invalid HTML tag found in project_id: {id}. Escaping HTML..."
+                )
+                record["attributes"]["project_status_update"] = html.escape(
+                    status_update
+                )
```

Comment:
TIL about the built-in `html` library 👍 

since this ends up loading html tags as text into the status update, i wanted to offer one alternative, which would be to use bs4 to grab the text content instead.

```shell
soup.get_text()
```

in the case of our busted html that spits out:
```
'Bluebeam session for design review by PDD PM (by Friday, 9/20)Cesar Chavez Blank Out SignsSession ID: 187-616-135Session URL: https://studio.bluebeam.com/hyperlink.html?link=studio.bluebeam.com/sessions/187-616-13 
```

i'm not sure how reliable this—maybe worth punting it to another issue.

Reply:
i like this idea a lot. Feels better than outputting HTML tags and will look nicer. Thanks for finding this method!